analyzers/src/SonarAnalyzer.Common/Helpers/FilesToAnalyzeStorage.cs

```diff
@@ -0,0 +1,48 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace SonarAnalyzer.Helpers
+{
+    public class FilesToAnalyzeStorage
+    {
+        private readonly Lazy<IEnumerable<string>> filestoAnalyze;
+
+        [ExcludeFromCodeCoverage]
+        public FilesToAnalyzeStorage(string filePath) : this(filePath, new FilesToAnalyzeRetriever()) { }
+
+        internal FilesToAnalyzeStorage(string filePath, IFilesToAnalyzeRetriever filesToAnalyzeRetriever)
+        {
+            filestoAnalyze = new Lazy<IEnumerable<string>>(() => filesToAnalyzeRetriever.RetrieveFilesToAnalyze(filePath));
+        }
```

Comment:
I believe this could be simplified a lot with removing the interface and the other class.
Let's don't use Lazy here and directly load all values. The state&caching will be handled by the user (Andrei's PR)

Reply:
I think the main idea I have is not to do the I/O operation in the constructor, because operations that can throw exceptions should be avoided in the constructor.

Reply:
WDYT about removing the interface and the other class and leaving the Lazy here. I would say it can do ho harm anyway. And remove the lazy stuff eventually when the caching is implemented in a different place.

Reply:
My thoughts about the situation:
* I'm 100% sure that caching will be implemented else where (#4119, `ProjectConfig` will be properly cached and will hold a lazy reference to this class).
* Throwing exception from constructor is acceptable in our scenarios I think (easy to understand, easy to debug on current and future version). Same as throwing `ArgumentNullException` for example.
* It's also subject to [RFC-1925 6a)](https://tools.ietf.org/html/rfc1925) (If you don't know, the RFC itself is one of April's fool day jokes, but the content is a valid warning).
* The intended usage in this and the future scenario will be create&use directly. So throwing an exception from constructor is not a problem. 

I'm for removing the lazyness from this.

Reply:
Indeed if we implement #4119  it does not make sense to have a lazy implementation here. I`ll remove it as it looks like we will have the time to implement #4119